src/vs/workbench/parts/extensions/electron-browser/extensionsViewlet.ts

```diff
@@ -463,11 +475,16 @@ export class ExtensionsViewlet extends ViewContainerViewlet implements IExtensio
 
 	protected onDidAddViews(added: IAddedViewDescriptorRef[]): ViewletPanel[] {
 		const addedViews = super.onDidAddViews(added);
-		this.progress(Promise.all(addedViews.map(addedView => {
-			(<ExtensionsListView>addedView).show(this.normalizedQuery()).then(model => {
-				this.alertSearchResult(model.length, addedView.id);
-			});
-		})));
+
+		this.searchPromise = createCancelablePromise(token => {
```

Comment:
We should probably track this separately and not override the `searchPromise` variable.

Maybe have `searchPromiseForExistingViews` and `searchPromiseForNewViews` ?

Reply:
When you are typing one letter `doSearch` and `onDidAddViews` methods are called and they are searching over the same view.

Reply:
Say you have `viewA` open and you have typed some text in the search box.
Based on the text, new views may be added which triggers search in the new views.
Based on current code, `searchPromise` will track the search on the new views.
No-one is tracking the search on the original view.

Say more text gets typed. Now with the current code, we will only end up cancelling the search in the new views and not the original view. So the race condition can still happen for the original view.

Reply:
I propose to replace `searchPromise: CancelablePromise ` with `searchPromises: Map<ViewletPanel, CancelablePromise<any>>`, so we can create `CancelablePromise` for every `ViewletPanel` and cancel search only for those panels that are in search condtion now.

Reply:
Proposed solution code https://github.com/grunxen/vscode/commit/8b7caa409a59c87dceabd3de4e07fc178604035e

Reply:
That looks good

Reply:
PR updated